### PR TITLE
proto: move keys to oneof

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

- define proto oneof for keys (only ed25519)

- remove tm2pb and pb2tm util functions for keys and define them in the crypto/encoding package

______

For contributor use:

- [ ] Wrote tests
- [ ] Updated CHANGELOG_PENDING.md
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by marbar3778** [comment]
need to split this PR to help unblock some IBC work, will be putting the proto types into a different PR

**Comment by marbar3778** [comment]
blocked on #4661

File: types/evidence.go [hunk 1]
```diff
@@ -19,12 +19,12 @@ import (
...
 const (
 	// MaxEvidenceBytes is a maximum size of any evidence (including amino overhead).
-	MaxEvidenceBytes        int64 = 436
-	ValidatorsHashField           = "ValidatorsHash"
-	NextValidatorsHashField       = "NextValidatorsHash"
-	ConsensusHashField            = "ConsensusHash"
-	AppHashField                  = "AppHash"
-	LastResultsHashField          = "LastResultsHash"
+	MaxEvidenceBytes        int64  = 436
+	ValidatorsHashField     string = "ValidatorsHash"
```
Comment (melekes) [inline types/evidence.go:23]: ?? why add the type

File: types/vote_set.go [hunk 2]
```diff
@@ -163,9 +163,7 @@ func (voteSet *VoteSet) addVote(vote *Vote) (added bool, err error) {
 	blockKey := vote.BlockID.Key()
 
 	// Ensure that validator index was set
-	if valIndex < 0 {
-		return false, errors.Wrap(ErrVoteInvalidValidatorIndex, "Index < 0")
```
Comment (melekes) [inline types/vote_set.go:167]: should we remove ErrVoteInvalidValidatorIndex too? or it's used somewhere else?

File: types/evidence.go (hunk 1 above)
Comment (marbar3778) [reply types/evidence.go:23]: linter was complaining that maxEvidenceBytes is the only one with type so added them below as well.

File: types/vote_set.go (hunk 2 above)
Comment (marbar3778) [reply types/vote_set.go:167]: its still used elsewhere

**Pull request merged**